### PR TITLE
Create cubemap from file

### PR DESCRIPTION
**Pull request by johannesugb**
@stf976

### CODE REVIEW

**Comment by johannesugb** [comment]
@stf976 not sure why all the compile checks are failing. Could you take a look?

**Comment by stf976** [comment]
> 
> 
> @stf976 not sure why all the compile checks are failing. Could you take a look?

msvc failed because it cannot find Vulkan:

```
-- Could NOT find Vulkan (missing: Vulkan_LIBRARY Vulkan_INCLUDE_DIR) 
-- Configuring done
CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
Please set them or make sure they are set and tested correctly in the CMake files:
D:/a/Auto-Vk/Auto-Vk/Vulkan_INCLUDE_DIR
   used as include directory in directory D:/a/Auto-Vk/Auto-Vk
```

clang and gcc failed because in more recent vulkan_core.h headers, VK_NULL_HANDLE  on some platforms is defined as a nullptr, which is of std::nullptr_t type. Older versions (e.g. 135 which I am using) simply define VK_NULL_HANDLE as integer 0. The error messages for gcc are:

 /home/runner/work/Auto-Vk/Auto-Vk/src/avk.cpp:1614:22: error: cannot convert ‘std::nullptr_t’ to ‘vk::DeviceAddress’ {aka ‘long unsigned int’}
 1614 |    .setDeviceAddress(VK_NULL_HANDLE); // TODO: support this (deviceAddress is the device address requested for the acceleration structure if the rayTracingAccelerationStructureCaptureReplay feature is being used.)

/home/runner/work/Auto-Vk/Auto-Vk/src/avk.cpp:1961:22: error: cannot convert ‘std::nullptr_t’ to ‘vk::DeviceAddress’ {aka ‘long unsigned int’}
 1961 |    .setDeviceAddress(VK_NULL_HANDLE);

**Review by johannesugb — COMMENTED** [review]
> clang and gcc failed because in more recent vulkan_core.h headers, VK_NULL_HANDLE on some platforms is defined as a nullptr, which is of std::nullptr_t type. Older versions (e.g. 135 which I am using) simply define VK_NULL_HANDLE as integer 0. The error messages for gcc are:

Okay. Thanks. Pushed a version that should compile against SDK 1.2.176.1

@stf976  Could you rebase on the most up-to-date `master` branch's version and create a new PR? 

Hopefully the whitespace-issue is gone afterwards, too; Actually I can't see which of the commits above has added them. Maybe it is because it was branched-off from a version that still had the whitespace-issue?

**Comment by johannesugb** [comment]
> Hopefully the whitespace-issue is gone afterwards, too; Actually I can't see which of the commits above has added them. 

Hmm, obviously the whitespaces have been added via the last commit: 4c96f604adb53f1ddd8c1cc6a63cf9aed6297525

I.e. you'd have to get rid of that last commit. Two strategies come to mind:

1. Reset the head pointer of this branch to one commit earlier, and then rebase on the current `master`
2. Make a new branch by branching-off from the current `master`, and `git cherry-pick` one commit after another in the right order, leaving out the problematic commit. 

Then make a new PR!

**Comment by stf976** [comment]
That comment is from the master revision I created my branch off. I'll update it with the next merge.

**Comment by stf976** [comment]
> 
> 
> Nope `eBottomOfPipe` is unsuitable for creating **memory**-barriers. A specific pipeline stage must be used so that a proper memory barrier can be established. BLITs are transfer operations. Therefore, `eTransfer` is the right pipeline stage.

That comment is from the master revision I created my branch off. I'll update it with the next merge.

File: src/avk.cpp [hunk 1]
```diff
@@ -1454,7 +1454,17 @@ namespace avk
 				return vk::ImageViewType::e1D;
 			}
 		case vk::ImageType::e2D:
-			if (info.arrayLayers > 1) {
+			if (info.flags & vk::ImageCreateFlagBits::eCubeCompatible)
+			{
+				// TODO: should a cubemap array with only 1 cubemap be supported?
```
[the remark this answers is not in the document]
Comment (stf976) [reply src/avk.cpp:1459]: So far I don't see a need to support it, unless you know of a use case where this might be handy.

**Comment by stf976** [comment]
> 
> 
> > Hopefully the whitespace-issue is gone afterwards, too; Actually I can't see which of the commits above has added them.
> 
> Hmm, obviously the whitespaces have been added via the last commit: [4c96f60](https://github.com/cg-tuwien/Auto-Vk/commit/4c96f604adb53f1ddd8c1cc6a63cf9aed6297525)
> 
> I.e. you'd have to get rid of that last commit. Two strategies come to mind:
> 
>     1. Reset the head pointer of this branch to one commit earlier, and then rebase on the current `master`
> 
>     2. Make a new branch by branching-off from the current `master`, and `git cherry-pick` one commit after another in the right order, leaving out the problematic commit.
> 
> 
> Then make a new PR!

The extra whitespace was already in master when I created my new branch. They were later on removed by [3fac798](https://github.com/cg-tuwien/Auto-Vk/commit/3fac798c9d8f9dad2ae6e280f0e7a7c056e21993). The commit  [4c96f60](https://github.com/cg-tuwien/Auto-Vk/commit/4c96f604adb53f1ddd8c1cc6a63cf9aed6297525) did not change anything about the whitespace in question.

Merging by hand would have been a chore as the file in question has more 8k lines. Fortunately my changes were small enough and don't interfere with others, so I created a new branch and re-applied my code (see new pull request).

This is a prime example why whitespace cleanups are generally a Very Bad Idea (tm). No contributor or reviewer should have to deal with these massive changesets that are mixed with actual code. It not only makes merging branches and reviews unnecessarily difficult, but also the comparison of different revisions later on.

If consistent trailing white space is a real practical concern, I would strongly recommend to use a pre-commit hook on git, as e.g. mentioned here: 
https://stackoverflow.com/questions/591923/make-git-automatically-remove-trailing-white-space-before-committing 
Anything else is just a waste of time for everybody involved.

**Comment by johannesugb** [comment]
> Merging by hand would have been a chore as the file in question has more 8k lines. Fortunately my changes were small enough and don't interfere with others, so I created a new branch and re-applied my code (see new pull request).

Yeah, I know. Sorry that this has been such a mess. I don't really know where the problematic whitespaces came from in the first place.

> This is a prime example why whitespace cleanups are generally a Very Bad Idea (tm). No contributor or reviewer should have to deal with these massive changesets that are mixed with actual code. It not only makes merging branches and reviews unnecessarily difficult, but also the comparison of different revisions later on.

I totally agree.

> If consistent trailing white space is a real practical concern, I would strongly recommend to use a pre-commit hook on git, as e.g. mentioned here:
> https://stackoverflow.com/questions/591923/make-git-automatically-remove-trailing-white-space-before-committing
> Anything else is just a waste of time for everybody involved.

That's a good point. I do not have a lot of experience with setting up such hooks, but maybe someone else has and is willing to iplement it. 
May I ask you to create an Issue, suggesting to install such a hook?

**Pull request closed without merging**